File: Leetcode/Python3/45. Jump Game II/solution.py
# https://leetcode.com/problems/jump-game-ii/

# Ref: https://leetcode.com/problems/jump-game-ii/discuss/170518/8-Lines-in-Python!-Easiest-Solution!
class Solution:
    def jump(self, nums: List[int]) -> int:
        length = len(nums)
        if length <= 1:
            return 0
        left, right = 0, nums[0]
        times = 1
        while right < length - 1:
            times += 1
            left, right = right, max(i + nums[i] for i in range(left, right + 1))
        return times

                
            

File: Leetcode/C++/45. Jump Game II/solution.cpp
// https://leetcode.com/problems/jump-game-ii/

#include <algorithm>
#include <vector>

using namespace std;

class Solution
{
public:
    int jump(vector<int> &nums)
    {
        int time = 1, left = 0, right = nums[0];
        if (nums.size() <= 1)
        {
            return 0;
        }
        while (right < nums.size() - 1)
        {
            time += 1;
            int maxJump = 0;
            for (int i = left; i <= right; i++)
            {
                maxJump = max(maxJump, i + nums[i]);
            }
            left = right;
            right = maxJump;
        }
        return time;
    }
};
